### ISSUE TITLE
Bug: [no-unecessary-condition] false positive on untyped variable that may be unassigned

### ISSUE THREAD

**Issue by jedwards1211** [typescript-eslint/typescript-eslint#10983]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=KYDwDg9gTgLgBDAnmYcAKUIFsCWBnYAdRxgAsAlYPCAGwDdgo8AeAFQD44BeOAbwCg4cMJlwEAXOlH5gbdoLhQqtBpIAUdAIY0ArsEms4AHynYZAGRwBrWRwCU3TnQg4AJgqUArYAGMY6pU1qADsAfklNYMQHLicXdwBffn4AegAqNME0uABVAigaSNc4HCwwGmAsYGCYTRgcCGC4CAAzOABtDDMCADoAdxIKZXpGPABdXgABGhxgqzhSGBgwPHEUlNdgBhoIFCgerAgALxwaQp7oAHMU6oBaHIBlDYgfPBTCYAAjFIApTS0Hj4oDgwDAUpQWoxqj5gCkAOI7T7aAD6AHlPt4-G8umJYQMyJRqCMmAkelk4KjgjC4H1UDowJcoJpNggIHAAHIQFkAJm5ABoaagfJE4Hh8T5SKyEKRUJ8dKd6sEyXA0il%2BKBILA4C0dFT6o0aYNCSpRnI1HZJDiZMQCcMGEw5OrwNB4Dq9Q0mvihkT7SwOGoFGR8JbpAQAMKNPAwKA6PzQfgW0y4m3ek0OjhOzWu3V%2BD2G20%2B03%2BwOkYNJmQR4JRmNxqDGOC6zYtWbAdyJq0EFPG4l%2BzgCIQ%2BSPwDvASvV2MwaDcaX4YwmEcKCrwJSFgVeXwwBSDqvwETdVA8YLAPrl8NDmuTqBmtQrtfATwxPsKIQrk3TlfPxT3jfv%2B8KBJ2AoOBtGoACEr4jHOcDgd%2BfgOP2QjSpgJ5HieACiUCYFAAaIYhADkEEMM0dbrn4NJBA2wRNi2xQykoApBngoqkBAOg0MUR72gsmhgCgwR4Z%2BgFCEkL7ADAOhQE0vDCKGwB3quX6YvASRJEAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
export type PromiseWithResolvers<T> = {
  promise: Promise<T>
  resolve: (value: T | PromiseLike<T>) => void
  reject: (reason?: any) => void
}

/**
 * Userland implementation of [Promise.withResolvers]{@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers}.
 * Once we upgrade to Node 22, we can switch to the builtin.
 */
export function withResolvers<T>(): PromiseWithResolvers<T>
export function withResolvers<T>(
  this: PromiseConstructor
): PromiseWithResolvers<T>
export function withResolvers<T>(
  this: PromiseConstructor | undefined
): PromiseWithResolvers<T> {
  const PromiseConstructor = this || Promise
  let resolve, reject
  const promise = new PromiseConstructor<T>((res, rej) => {
    resolve = res
    reject = rej
  })
  if (!resolve || !reject) {
    throw new Error(
      'resolve or reject was undefined here, this should never happen'
    )
  }
  return { promise, resolve, reject }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No error from ESLint

### Actual Result

Unnecessary conditional, value is always truthy. 24:7 - 24:15
Unnecessary conditional, value is always truthy. 24:19 - 24:26

### Additional Info

This is a minor issue since I might as well just add type annotations.  But it seems concerning that the inferred type `typescript-eslint` gets at this position is defined when TS considers the variables possibly undefined (if I remove the type guard, I get TS errors on the `return` statement since `resolve`/`reject` could be `undefined`).

**Comment by kirkwaiblinger** [comment]
Hi @jedwards1211!

> ... the inferred type `typescript-eslint` gets at this position is defined...

It's the opposite, actually. The type of the variable is given by TS as `undefined`, and the rule is indicating that `!resolve` is therefore always `true` (which, you might feel is even more wrong 😀). This is [one of the classic limitations](https://typescript-eslint.io/rules/no-unnecessary-condition/#values-modified-within-function-calls)... 

I think the "untyped" behavior (often referred to as an ["evolving" type](https://effectivetypescript.com/2020/03/09/evolving-any/)) is kind of a red herring here, since really it's just TS's CFA not being aware that the `resolve = res` assignment in the callback is immediately executed.

**Comment by kirkwaiblinger** [comment]
Also, fun fact, the reason that the code _works_ is actually not because TS assesses that `resolve` has a valid type. It's that after the null check, there's nothing left that it could be, so it's narrowed to `never`, which is assignable to anything ([see playground](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=KYDwDg9gTgLgBDAnmYcAKUIFsCWBnYAdRxgAsAlYPCAGwDdgo8AeAFQD44BeOAbwCg4cMJlwEAXOlH5gbdgG5BcKFVoNJAOwCuWAEaN5cAPRG4zALTm4SlQCtgAYxiS8MKDg0BzQybOXrAL78oJCwcABmWhpOOBAacADuJBSq9IwsHAAUSmT4khjYMgDCca5QWk7QcAA%2BcFEAJsDhHsD1-ACU%2BdIExGSU1GlMcnxKDqXwBWLAJRplFTBVPLl4NbWTMko0wPAqAwwANMrA9k6j48LdqDwawAlShQQzc5VQcpmZu4d27dycAkJCXZqK5HPBKQHHRzwHh2JQBdpKHDhOCZACEQLSqzg6MhTh%2B-wBZEwdxudwAolBMFBsgCAQByDEMOBVOxQxIAQxWDSaLXqcFIjGAh2WcDwpAgWhofJuDCg-PZYBQGjp4LgCKEQRsqQYiiEJgAegB%2BIRak4wXXGIxG-hKXyM1D4IQaCDwdlwbR6RhwdkaPmspxwR3O12itweTyorUwLRQeK8C4PIWg4FfXHwIIBIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRQA7iKaRwYAL4hNQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false)).

I just bring this up to say that I think the rule is acting correctly by taking TS at its word here, and it's really just TS's behavior that's somewhat surprising!

**Comment by jedwards1211** [comment]
Oh I see.

I don't suppose y'all would be open to a rule option that makes it ignore conditions containing variables that are modified within functions?

**Comment by bradzacher** [comment]
https://typescript-eslint.io/rules/no-unnecessary-condition#when-not-to-use-it

This is expected and noted in the docs.
